### DISCUSSION TITLE
A fast alternative to the picom X11 compositor - fastcompmgr

### DISCUSSION THREAD

**Discussion by tycho-kirchner** [i3/i3#6039] [Show and tell]
Since the early days of compton, every new release made my desktop feel somewhat slower; picom, unfortunately, continues this sad tradition. That's why I travelled back in time to an early version of compton (which added shadows on argb windows), cherry picked some later compton commits to get rid of spurious segfaults and memleaks, and made that version even faster, based on profiling. Since then moving and resizing windows or scrolling complex web pages is finally smooth again. You may want check it out here:  [https://github.com/tycho-kirchner/fastcompmgr](https://github.com/tycho-kirchner/fastcompmgr).

CPU usages by compositor:

| Compositor    | move  | resize  | scroll |
| ------------- | ----- | ------- | ------ |
| fastcompmgr   | 6.7%  | 4.4%    | 1.5%   |
| xcompmgr      | 7.8%  | 4.9%    | 1.6%   |
| compton       | 26.4% | 6.8%    | 17.1%  |
| picom         | 29.3% | 8.1%    | 23.1%  |

**Comment by Samueru-sama** [comment]
This looks interesting but how does one get rid of the screen tearing when using this compositor?

**Comment by tycho-kirchner** [comment]
Oddly, when visiting [this site](http://vsynctester.com/), I experience an overall better performance with fastcompmgr as opposed to picom. But to answer your question, fastcompmgr, just like the original xcompmgr, currently has no specific means to get rid of tearing. If people continue on requesting that feature I might do so in the future 🙂.

**Comment by Samueru-sama** [comment]
> Oddly, when visiting [this site](http://vsynctester.com/), I experience an overall better performance with fastcompmgr as opposed to picom. But to answer your question, fastcompmgr, just like the original xcompmgr, currently has no specific means to get rid of tearing. If people continue on requesting that feature I might do so in the future 🙂.

I was able to get rid of the tearing by launching xorg with the `--set TearFree on` flag. 

Now I can use this compositor, here is a benchmark showing the cpu usage while moving a window around:

![image](https://github.com/i3/i3/assets/36420837/c024ffe0-c2d2-4e7f-9b13-787327ffa5a8)

If the test was right that would be a 20% reduction in cpu usage between the two, interesting.

**Comment by tycho-kirchner** [comment]
Ah, ok, so you mean you have an intel graphics card and use intel's old xserver-xorg-video-intel driver? Is this `--set TearFree` a commandline option for Xorg? I thought the only way to configure this was via `/etc/X11/xorg.conf.d/20_my_intel.conf`....
Btw, did you also enable `dri3` (see below)? On my machine I get less tearing using this option (but nowadays I use the modesetting driver anyway).
~~~
Section "Module"
   Load "dri3"
EndSection

Section "Device"
   Identifier  "Intel Graphics"
   Driver      "intel"
# DRI affects only intel
  Option "DRI" "3"
EndSection


~~~

**Reply by Samueru-sama** [reply]
@tycho-kirchner I have AMD, what I did is detailed here in the archwiki: https://wiki.archlinux.org/title/AMDGPU#Tear_free_rendering

I already use a script that sets each of my displays to a custom refresh rate, so I just had to add that option to it. 

However I just noticed a bug with fastcompmgr, I cannot take screenshots with `xfce4-screenshooter` that I use on i3wm, as the applications turn 100% transparent when using it and I cannot select anything as result.

**Reply by tycho-kirchner** [reply]
Somehow, I lost you comment back then. This should however have been fixed meanwhile 😇.
